### PR TITLE
Relation CSE

### LINKED ISSUE #5674: Redundant join removal fails when there is a filter on a join column.
Original question from @philip-stoev:
Contrast
```
materialize=> explain SELECT * FROM orders AS outer_orders WHERE o_orderkey IN
materialize-> ( SELECT o_orderkey FROM orders AS inner_orders);
             Optimized Plan             
----------------------------------------
 %0 =                                  +
 | Get materialize.public.orders (u439)+
 | ArrangeBy (#0)                      +
                                       +
 %1 =                                  +
 | Get materialize.public.orders (u439)+
 | Filter !(isnull(#0))                +
 | Distinct group=(#0)                 +
 | ArrangeBy (#0)                      +
                                       +
 %2 =                                  +
 | Join %0 %1 (= #0 #9)                +
 | | implementation = DeltaQuery       +
 | |   delta %0 %1.(#0)                +
 | |   delta %1 %0.(#0)                +
 | | demand = (#0..#8)                 +
 | Project (#0..#8)  
```
with 
```
materialize=> explain SELECT * FROM orders AS outer_orders WHERE o_orderkey IN 
materialize-> ( SELECT o_orderkey FROM orders AS inner_orders WHERE o_orderkey  = 16);
             Optimized Plan             
----------------------------------------
 %0 =                                  +
 | Get materialize.public.orders (u439)+
 | ArrangeBy (#0)                      +
                                       +
 %1 =                                  +
 | Get materialize.public.orders (u439)+
 | Filter (#0 = 16)                    +
 | Distinct group=(16)                 +
 | ArrangeBy (#0)                      +
                                       +
 %2 =                                  +
 | Get materialize.public.orders (u439)+
 | ArrangeBy (#0)                      +
                                       +
 %3 =                                  +
 | Join %1 %2 (= #0 #1)                +
 | | implementation = DeltaQuery       +
 | |   delta %1 %2.(#0)                +
 | |   delta %2 %1.(#0)                +
 | | demand = (#0)                     +
 | Filter (#0 = 16)                    +
 | Distinct group=(16)                 +
 | ArrangeBy (#0)                      +
                                       +
 %4 =                                  +
 | Join %0 %3 (= #0 #9)                +
 | | implementation = DeltaQuery       +
 | |   delta %0 %3.(#0)                +
 | |   delta %3 %0.(#0)                +
 | | demand = (#0..#8)                 +
 | Project (#0..#8)                    +
```

The second query just has an extra filter compared to the first query. But why does the plan for the second query have an extra join?

The short answer is the RedundantJoin transform can recognize that
```
%0 = Let l0 =
| Get ? (u1)
| Distinct group=(#0)

%1 =
| Get ? (u1)

%2 =
| Join %0 %1 (= #0 #1)
| | implementation = Unimplemented
```
is equivalent to `Get ? (u1)`.

But it cannot recognize that 
```
%0 = Let l0 =
| Get ? (u1)
| Distinct group=(#0)

%1 =
| Get %0 (l0)
| Filter (#0 = 16)

%2 =
| Get ? (u1)
| Filter (#0 = 16)

%3 =
| Join %1 %2 (= #0 #1)
| | implementation = Unimplemented
``` 
is equivalent to 
```
| Get ? (u1)
| Filter (#0 = 16)
```

### LINKED ISSUE #6748: Predicate pushdown can prevent redundant join from doing its job
`create table t1(f1 int, f2 int, f3 int);`

In the query below, our `RedundantJoin` optimization successfully removes the join:
```
materialize=> explain 
select * 
from t1 a1, (select f1, f2 from t1 group by f1, f2) a2 
where (a1.f1 = a2.f1 or (a1.f1 is null and a2.f1 is null))
  and (a1.f2 = a2.f2 or (a1.f2 is null and a2.f2 is null));
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u53)+
 | Project (#0..#2, #0, #1)       +
```

However, if you tweak the query slightly, `RedundantJoin` becomes unable to remove the join because a filter has been pushed to each input of the join. 
```
materialize=> explain 
select * 
from t1 a1, (select f1, f2 from t1 group by f1, f2) a2 
where (a1.f1 = a2.f1 or (a1.f1 is null and a2.f1 is null))
  and a1.f2 = a2.f2;
                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u53)               +
 | Filter !(isnull(#1))                          +
                                                 +
 %1 =                                            +
 | Get materialize.public.t1 (u53)               +
 | Filter !(isnull(#1))                          +
 | Distinct group=(#0, #1)                       +
 | ArrangeBy (#0, #1)                            +
                                                 +
 %2 =                                            +
 | Join %0 %1 (= #0 #3) (= #1 #4)                +
 | | implementation = Differential %0 %1.(#0, #1)+
 | | demand = (#0..#2)                           +
 | Project (#0..#2, #0, #1)                      +
```

This issue cannot be solved by running `RedundantJoin` before `PredicatePushdown`. `RedundantJoin` cannot remove the join until it knows the equivalence classes of the join. `PredicatePushdown` sets the equivalence classes of the join at the same time that it pushes the filter down to both inputs of the join.

### LINKED ISSUE #7046: [optimization] Introduce RelationExpr CSE
Our analysis will inline `Let` expressions, but does no hard work to identify opportunities to create `Let` expressions that re-use expressions. Re-using expressions would result in fewer dataflow fragments, which results in less computation and less state maintained, at least from those fragements.

The downside of `Let` bindings is that we lose the opportunity to optimize expressions independently: we can only push down maps, filters, projections, etc if all of the `Get` uses have the same things to push down. At the moment, the `render` layer does this for us, but with #7001 we will not be doing that. If nothing else, we should re-introduce a CSE optimization in the final steps of dataflow optimization, to avoid regressions there.

### PR DESCRIPTION
**Pull request by frankmcsherry**
Common subexpression elimination for Relations.

This is currently "sloppy" and relies on `InlineLet` to clean up after it. It could do more work, but I didn't want to encode our Let inlining rules in two places. It may be smart to do some of this anyhow, to minimize the volume of output (which otherwise has a Let binding for each distinct expression node).

CSE has the potential defect that it prevents specialization. E.g. a join of two tables that is used twice, with different filters, may result in a CSE'd cross join and then two filters applied to Get expressions. Not great, and not prevented here. The changes in plans do not look bad though. We could avoid re-using joins that contain cross-join substructure, for example.

I have put the analysis in two random places in `transform/lib.rs`, and could use some input about the best places to put them to be fully effective. Only placing them in the first location does not result in complete optimization, for reasons that were not obvious at time.

Fixes #6748 
Fixes #5674 
Fixes #7046

### CODE REVIEW

**Comment by frankmcsherry** [comment]
@philip-stoev this has the potential to (and indeed has) mess around with a lot of query plans. It has thusfar only done ok things that I've seen, but it could make some things works (e.g. introduce cross joins as mentioned in the desc). If you would like to test it out, that would be great!

**Comment by philip-stoev** [comment]
@frankmcsherry should I expect that a `SELECT * FROM t1, t1``` will be recognized and folded to use a single `Get` operator? Currently, it does not seem to do that, it only does it for more complex queries than that.

**Comment by frankmcsherry** [comment]
It should do that if `t1` is anything other than a `Get`. It will not replace two `Get` operators with two new `Get` operators that reference a local binding of the common `Get` operator.

**Comment by philip-stoev** [comment]
Item No1 . The CTE detection does not kick in this case:

```
materialize=> EXPLAIN SELECT * FROM t1 AS a1 JOIN t1 AS a2 USING (f1) WHERE a1.f1 = 1 AND a1.f2 = 2;
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.t1 (u403)        +
 | Filter (#0 = 1), (#1 = 2)               +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.t1 (u403)        +
 | Filter (#0 = 1)                         +
                                           +
 %2 =                                      +
 | Join %0 %1                              +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0, #1, #3)                 +
 | Project (#0, #1, #3)                    +
```

The ```Filter (#0 = 1)```  is applied twice, on two separate `Get`s

**Review by philip-stoev — APPROVED** [review]
I created and pushed a dedicated SLT test for this and seems that (almost) all of the CSEs I could think of are indeed being detected as such.

A randomized workload reported no queries where the performance is now worse.

I was unable to observe any cross joins of the type mentioned in the description with either manually-written or randomized queries.

File: test/sqllogictest/chbench.slt [hunk 1]
```diff
@@ -300,24 +300,25 @@ AND i_id = m_i_id
 AND s_quantity = m_s_quantity
 ORDER BY n_name, su_name, i_id
 ----
-%0 =
-| Get materialize.public.item (u24)
+%0 = Let l0 =
+| Get materialize.public.region (u37)
+| Filter "^EUROP.*$" ~(#1)
```
[the remark this answers is not in the document]
Comment (frankmcsherry) [reply test/sqllogictest/chbench.slt:305]: I could imagine that the Let/Get pair gets in the way of predicate re-hoisting.

**Comment by frankmcsherry** [comment]
Thanks folks. I filed #7825 about detecting even finer subexpressions (those that would require breaking apart expression stages). I'm going to look in to one instance of `Filter` that Andi saw that idk what is going on (we cannot remove a `Filter` for some reason), but I'm inclined to merge now and follow with that.

**Pull request merged**